test/gateway/retry-failed-services.js

```diff
@@ -0,0 +1,513 @@
+        }
+      ],
+      retryServicesCount: 10,
+      retryServicesInterval: 2000
+    }
+  })
+
+  await app.ready()
+
+  const query = `
+    query {
+      user: me {
+        id
+        name
+        posts(count: 1) {
+          pid
+        }
+      }
+    }`
+
+  const res = await app.inject({
+    method: 'POST',
+    headers: { 'content-type': 'application/json' },
+    url: '/graphql',
+    body: JSON.stringify({ query })
+  })
+
+  t.same(JSON.parse(res.body), {
+    errors: [
+      {
+        message: 'Cannot query field "posts" on type "User".',
+        locations: [{ line: 6, column: 9 }]
+      }
+    ],
+    data: null
+  })
+
+  for (let i = 0; i < 10; i++) {
+    await clock.tickAsync(1000)
+  }
+
+  const res1 = await app.inject({
+    method: 'POST',
+    headers: { 'content-type': 'application/json' },
+    url: '/graphql',
+    body: JSON.stringify({ query })
+  })
+
+  t.same(JSON.parse(res1.body), {
+    data: {
+      user: {
+        id: 'u1',
+        name: 'John',
+        posts: [
+          {
+            pid: 'p1'
+          }
+        ]
+      }
+    }
+  })
+})
+
+test('gateway - dont retry non-mandatory failed services on startup', async (t) => {
+  t.plan(2)
+  const clock = FakeTimers.install({
+    shouldAdvanceTime: true,
+    advanceTimeDelta: 50
+  })
+
+  const service1 = await createTestService(5001, userService.schema, userService.resolvers)
+
+  const app = Fastify()
+  t.teardown(async () => {
+    await app.close()
+    await service1.close()
+    clock.uninstall()
+  })
+
+  app.register(GQL, {
+    jit: 1,
+    gateway: {
+      services: [
+        {
+          name: 'user',
+          url: 'http://localhost:5001/graphql',
+          mandatory: false
+        },
+        {
+          name: 'post',
+          url: 'http://localhost:5002/graphql',
+          mandatory: false
+        }
+      ],
+      pollingInterval: 2000
+    }
+  })
+
+  await app.ready()
+
+  const query = `
+    query {
+      user: me {
+        id
+        name
+        posts(count: 1) {
+          pid
+        }
+      }
+    }`
+
+  const res = await app.inject({
+    method: 'POST',
+    headers: { 'content-type': 'application/json' },
+    url: '/graphql',
+    body: JSON.stringify({ query })
+  })
+
+  t.same(JSON.parse(res.body), {
+    errors: [
+      {
+        message: 'Cannot query field "posts" on type "User".',
+        locations: [{ line: 6, column: 9 }]
+      }
+    ],
+    data: null
+  })
+
+  await clock.tickAsync(10000)
+
+  const res1 = await app.inject({
+    method: 'POST',
+    headers: { 'content-type': 'application/json' },
+    url: '/graphql',
+    body: JSON.stringify({ query })
+  })
+
+  t.same(JSON.parse(res1.body), {
+    errors: [
+      {
+        message: 'Cannot query field "posts" on type "User".',
+        locations: [{ line: 6, column: 9 }]
+      }
+    ],
+    data: null
+  })
+})
+
+test('gateway - stop retrying after no. of retries exceeded', async (t) => {
```

Comment:
I guess this is the problematic test which is failing the CI at times

Reply:
From past experience, it could be related to the hardcoding of the port to 500(1|2) in the tests if it's failing intermittently in the CI. I would suggest passing `0` to `.listen()` if you can. If you need access to the port for use in the gateway config, you could listen straightaway and trigger errors dynamically using hooks (there are alternative methods, for sure! :) ). As an example:

```js
let shouldError = true
const service2 = await createTestService(postService.schema, postService.resolvers)
service2.addHook('onRequest', async () => {
  if (shouldError) {
    throw new Error('kaboom')
  }
})
await service2.listen(0)

// some work ...

shouldError = false
```

Reply:
Any luck on this? I'm available to help debug this, this week, if you need :)

Reply:
Thanks a lot @jonnydgreen  No luck so far. I tried the way you suggested but the tests are still failing. Seems the problem lies elsewhere. I am working on this today and will dig deeper. I will provide an update soon. Thanks!

Reply:
Awesome, sounds good!